### DISCUSSION TITLE
How to get a type alias from a class?

### DISCUSSION THREAD

**Discussion by shyakadavis** [typegoose/typegoose#898] [Q&A] [answered]
Hi.

Say I wanted to infer/retrieve a type alias from my class; how would I achieve that?

Context:

This is my class:

```ts
import {
	prop,
	getModelForClass,
	post,
	ModelOptions,
	Severity,
} from '@typegoose/typegoose';
import mongoose from 'mongoose';

@post<LoanClass>('save', function (doc) {
	if (doc) {
		doc.id = doc._id.toString();
		doc._id = doc.id;
	}
})
@post<LoanClass[]>(/^find/, function (docs) {
	// @ts-ignore
	if (this.op === 'find') {
		docs.forEach((doc) => {
			doc.id = doc._id.toString();
			doc._id = doc.id;
		});
	}
})
@ModelOptions({
	schemaOptions: {
		timestamps: true,
		collection: 'loans',
	},
	options: {
		allowMixed: Severity.ALLOW,
	},
})
class LoanClass {
	@prop({ required: true })
	lenderAddress: string;

	@prop({ required: true })
	loanToken: string;

	@prop({ required: true })
	loanAmount: string;

	@prop({ required: true })
	collateralToken: string;

	@prop({ required: true })
	collateralAmount: string;

	@prop({ required: true })
	liquidationThreshold: string;

	@prop({ required: true })
	initialThreshold: string;

	@prop({ required: true })
	loanRepayDeadline: Date;

	@prop({ required: true })
	loanRequestDeadline: Date;

	@prop({ required: true })
	interestRate: string;

	@prop({ required: true })
	managerNFT: string;

	@prop({ default: 'v1' })
	managerNFTVersion: string;

	_id: mongoose.Types.ObjectId | string;

	id: string;
}

const Loan = getModelForClass(LoanClass);
export { Loan, LoanClass };
```

And I want to be able to use the inferred type alias when populating an array like so:

```ts
let some_array: TLoan[]; // how can I get the type alias for this

some_array = [
{
	collateralAmount: '',
	collateralToken: '',
	loanAmount: '',
	loanToken: '',
	loanRepayDeadline: new Date(),
	initialThreshold: '',
	interestRate: '',
	liquidationThreshold: '',
	lenderAddress: '',
	loanRequestDeadline: new Date(),
	managerNFT: '',
	managerNFTVersion: '',
}
];
```

---

While I have you, do you mind having a look at that class definition, and telling me if I'm doing something the wrong way. I'm new to Typgoose, and navigating the docs is kinda hard. For example, understanding the `public` prefix is something that escapes me and its importance/significance thereof.

Thanks for the consideration.

**Comment by hasezoey** [comment, marked as answer]
> Say I wanted to infer/retrieve a type alias from my class; how would I achieve that?

i am not quite sure what you mean with "type alias", but if you are referring to having a type of the class that does not include functions, then maybe look at [`FilterOutFunctionKeys`](https://github.com/typegoose/typegoose/blob/766134b142fa132527e645df725164272e7d9887/src/types.ts#L722), otherwise please provide more specifics on what that type should be

**Comment by shyakadavis** [comment]
Hi, @hasezoey 

Thanks for the reply.

And yes, that was exactly what I wanted. i.e. `FilterOutFunctionKeys`.

Basically, what I wanted was just a way to type my function params, and different arrays but adhering to the class's shape.

Thanks a lot. 🙂 

Also, if it's no bother, do you mind having a look at that class definition. From first glance, does it look like I'm doing something wrong? For example, I see from the docs you use the `public` prefix; what does that do specifically, in terms of Typegoose, at least?

**Reply by hasezoey** [reply]
> Also, if it's no bother, do you mind having a look at that class definition. From first glance, does it look like I'm doing something wrong?

the only thing from the provided class i see that could be wrong is `allowMixed: Severity.ALLOW,`, it is not great to disable Mixed warnings for everything, if you have a property that is gonna be mixed and it is not accidental, then the prop option should be used instead [`@prop({ allowMixed: Severity.CHOOSE })`](https://typegoose.github.io/typegoose/docs/api/decorators/prop#allowmixed)

> For example, I see from the docs you use the public prefix; what does that do specifically, in terms of Typegoose, at least?

please read [typescript Member Visibility](https://www.typescriptlang.org/docs/handbook/2/classes.html#member-visibility)
as for why typegoose chooses this way even if not necessary, is because we (i) have gone for that style choice (for tests and documentation)
TL;DR: `public` along with `protected` and `private` are typescript member visibility (like in java, C# and other languages), and members are `public` by default - those things do not exist at runtime so they dont affect typegoose (they do affect the typescript types though) (and `private` is completely separate from ES private fields starting with `#`, which are unsupported by typegoose)
